.pre-commit-config.yaml

```diff
@@ -0,0 +1,35 @@
+repos:
+# basic pre-commit
+  - repo: https://github.com/pre-commit/pre-commit-hooks
+    rev: v4.1.0
+    hooks:
+      - id: trailing-whitespace
+      - id: end-of-file-fixer
+      - id: check-added-large-files
+      - id: check-yaml
+      - id: check-toml
+      - id: detect-private-key
+# sorting imports
+  - repo: https://github.com/pycqa/isort
+    rev: 5.10.1
+    hooks:
+      - id: isort
+        args: ["--profile", "black"]
+# syntax linting and formatting
+  - repo: https://github.com/myint/autoflake
+    rev: v1.4
+    hooks:
+      - id: autoflake
```

Comment:
Is this necessary?

Reply:
Or at least provide a way to opt out of this 'feature'...

Reply:
it removes unused imports and variables, I don't see any reason why it should be optional, if you don't agree with this I can remove it.

Reply:
Unused imports and variables can sometimes be intentional (although usual due to bad design). I agree that we should not generally allow these, but flake8 should catch those. I think it is the programmer's responsibility to check those issues when the pre-commit runs and the flake8 hook fails, and either fix it (in most cases) or discuss in the PR whether we have to special-case them.

Reply:
Code quality-wise it's redundant with flake8, and philosophically it delegates the act of consciously writing clean code to a bot that edits code behind the scene. But I'm likely being paranoid here, also curious how others think about this.

Reply:
In spirit, I am with you, but in practice, I think the more layers of defense against messy code, the better.